### PR TITLE
dwarf-fortress: 0.44.10 -> 0.44.11

### PR DESCRIPTION
**Pull request by numinit**
Update everything, including dfhack and dwarf-therapist.

###### Motivation for this change

New Dwarf Fortress version: http://www.bay12games.com/dwarves/index.html#2018-06-23

Note that this is my first nixpkgs pull request, let me know if I did anything terribly wrong.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/games/dwarf-fortress/wrapper/default.nix [hunk 1]
```diff
@@ -41,7 +43,11 @@ in
 stdenv.mkDerivation rec {
   name = "dwarf-fortress-${dwarf-fortress-original.dfVersion}";
 
-  compatible = lib.all (x: assert (x.dfVersion == dwarf-fortress-original.dfVersion); true) pkgs;
+  compatible = lib.all (x: x.dfVersion == dwarf-fortress-original.dfVersion) [
+    dwarf-fortress-original
+    dwarf-fortress-unfuck
+    dfhack
+  ];
```
Comment (numinit) [inline pkgs/games/dwarf-fortress/wrapper/default.nix]: Not sure if this is right but it was complaining about not being to find dfVersion in all of the arguments

File: pkgs/games/dwarf-fortress/soundsense.nix [hunk 2]
```diff
@@ -12,7 +12,7 @@
 
 stdenv.mkDerivation rec {
   version = "2016-1_196";
-  dfVersion = "0.44.10";
+  dfVersion = "0.44.11";
```
Comment (matthewbauer) [inline pkgs/games/dwarf-fortress/soundsense.nix]: You didn't update the hash.

File: pkgs/games/dwarf-fortress/wrapper/default.nix (hunk 1 above)
Comment (matthewbauer) [reply pkgs/games/dwarf-fortress/wrapper/default.nix]: Yeah probably there is a dfVersion somewhere that needs to be updated.

File: pkgs/games/dwarf-fortress/soundsense.nix (hunk 2 above)
Comment (numinit) [reply pkgs/games/dwarf-fortress/soundsense.nix]: Good catch, was wondering why I needed to do that. Thought it was just assurance that all the versions are the same (which they aren't between now and the time dfhack gets updated)

Comment (matthewbauer) [reply pkgs/games/dwarf-fortress/soundsense.nix]: Oh nevermind it wouldn't effect the hash. I thought this was "version"

Comment (numinit) [reply pkgs/games/dwarf-fortress/soundsense.nix]: Do you know why themes aren't installing, and is it for a similar reason?

I get "attribute 'tergel' missing" when trying to set `theme = "tergel"`.

Comment (matthewbauer) [reply pkgs/games/dwarf-fortress/soundsense.nix]: Oops i think that's broken. Could you try replacing this:

https://github.com/numinit/nixpkgs/blob/0bb87fdd3f7bc60102d60be9454ee62fb650a469/pkgs/games/dwarf-fortress/default.nix#L21-L24

with ```inherit themes;```?

Comment (numinit) [reply pkgs/games/dwarf-fortress/soundsense.nix]: Yep, that fixed it, will add it to this pull. Still getting a hang of Nix...

**Comment by numinit** [comment]
I see dfhack and therapist updated. Will add them to this pull.

**Comment by numinit** [comment]
Hmmm... 

```Initialized empty Git repository in /nix/store/98q0dd767bi3jbpwc6br9l039g5hli2b-source/.git/
remote: Counting objects: 860, done.
remote: Compressing objects: 100% (787/787), done.
remote: Total 860 (delta 65), reused 359 (delta 39), pack-reused 0
Receiving objects: 100% (860/860), 2.45 MiB | 13.47 MiB/s, done.
Resolving deltas: 100% (65/65), done.
unpacking sources
unpacking source archive /nix/store/gbihjsjgjb3nx6gj0a5rx3rdy5smcrl4-source
From https://github.com/DFHack/dfhack
 * tag               0.44.11-alpha1 -> FETCH_HEAD
source root is source
patching sources
configuring
fixing cmake files...
Switched to a new branch 'fetchgit'
cmake flags: -GNinja -DCMAKE_BUILD_TYPE=Release -DCMAKE_SKIP_BUILD_RPATH=ON -DCMAKE_INSTALL_INCLUDEDIR=/nix/store/k8k0bnf75pgwrzpap76bss3cybz95p5v-dwarf-therapist-original-40.1.0/include -DCMAKE_INSTALL_LIBDIR=/nix/store/k8k0bnf75pgwrzpap76bss3cybz95p5v-dwarf-therapist-original-40.1.0/lib -DCMAKE_INSTALL_NAME_DIR=/nix/store/k8k0bnf75pgwrzpap76bss3cybz95p5v-dwarf-therapist-original-40.1.0/lib -DCMAKE_POLICY_DEFAULT_CMP0025=NEW -DCMAKE_OSX_DEPLOYMENT_TARGET= -DCMAKE_OSX_SYSROOT= -DCMAKE_FIND_FRAMEWORK=last -DCMAKE_STRIP=/nix/store/4qvrxzxa535y8304mk195x50b6p9607d-binutils-2.30/bin/strip -DCMAKE_RANLIB=/nix/store/4qvrxzxa535y8304mk195x50b6p9607d-binutils-2.30/bin/ranlib -DCMAKE_AR=/nix/store/4qvrxzxa535y8304mk195x50b6p9607d-binutils-2.30/bin/ar -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ -DCMAKE_INSTALL_PREFIX=/nix/store/k8k0bnf75pgwrzpap76bss3cybz95p5v-dwarf-therapist-original-40.1.0
Submodule 'depends/clsocket' (https://github.com/DFHack/clsocket.git) registered for path 'depends/clsocket'
Submodule 'library/xml' (https://github.com/DFHack/df-structures.git) registered for path 'library/xml'
Submodule 'plugins/isoworld' (https://github.com/DFHack/isoworld.git) registered for path 'plugins/isoworld'
Submodule 'plugins/stonesense' (https://github.com/DFHack/stonesense.git) registered for path 'plugins/stonesense'
Submodule 'scripts2' (https://github.com/DFHack/scripts.git) registered for path 'scripts'
error: key does not contain a section: .url
```

#39956 maybe? None of the submodules have spaces though...

**Comment by numinit** [comment]
Updated my local nixpkgs checkout; issue seems to be resolved.

**Comment by numinit** [comment]
Still an issue with stonesense, looks like it might be missing a dependency:

```/tmp/nix-build-dfhack-0.44.11-alpha1.drv-0/source/plugins/stonesense/GUI.cpp:77:1: error: 'GLhandleARB' does not name a type; did you mean 'sighandler_t'?
 GLhandleARB tinter;
 ^~~~~~~~~~~
 sighandler_t
/tmp/nix-build-dfhack-0.44.11-alpha1.drv-0/source/plugins/stonesense/GUI.cpp:78:1: error: 'GLhandleARB' does not name a type; did you mean 'sighandler_t'?
 GLhandleARB tinter_shader;
 ^~~~~~~~~~~
 sighandler_t
make[2]: *** [plugins/stonesense/CMakeFiles/stonesense.dir/build.make:271: plugins/stonesense/CMakeFiles/stonesense.dir/GUI.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:4501: plugins/stonesense/CMakeFiles/stonesense.dir/all] Error 2
make: *** [Makefile:152: all] Error 2
builder for '/nix/store/mi2bwx708rvm3l6m43cj0fz2i6s48amr-dfhack-0.44.11-alpha1.drv' failed with exit code 2
```

**Comment by numinit** [comment]
Having several problems if I want to use dfhack:

First, it's missing a symbol which is apparently found in unfuck.

```Using /home/df/.local/share/df_linux as Dwarf Fortress overlay directory.
If you do any changes in it, don't forget to clean it when updating the game version!
We try to detect changes based on data directories being symbolic links -- keep this in mind.

/nix/store/f91i0ahbshkqpd619y7jpw866iafyghc-dwarf-fortress-env-0.44.11/libs/Dwarf_Fortress: symbol lookup error: /nix/store/f91i0ahbshkqpd619y7jpw866iafyghc-dwarf-fortress-env-0.44.11/hack/libdfhack.so: undefined symbol: _ZN8rendererD1Ev
```

If I add the following to `wrapper/dwarf-fortress-init.in`:

```sh
unfuck_dir="@unfuck@"
if [ ! -z "$unfuck_dir" ]; then
  export LD_LIBRARY_PATH="$unfuck_dir:$LD_LIBRARY_PATH"
  export LD_PRELOAD="$unfuck_dir/libgraphics.so:$LD_PRELOAD"
fi```

And the following to `wrapper/default.nix`:

```nix
    unfuck = if stdenv.isLinux then "${dfunfuck}/lib" else ""
```

Then this happens:

```mkdir: /nix/store/2kcrj1ksd2a14bm5sky182fv2xwfhfap-glibc-2.26-131/lib/libc.so.6: version `GLIBC_2.27' not found (required by /nix/store/ilbirzqw57lfhk6x8m53c6fa5ygll4kn-dwarf_fortress_unfuck-0.44.11/lib/libgraphics.so)
mkdir: /nix/store/2kcrj1ksd2a14bm5sky182fv2xwfhfap-glibc-2.26-131/lib/libc.so.6: version `GLIBC_2.27' not found (required by /nix/store/kxlczbgs4prbil5gsvicbrjs6f951fk9-libpulseaudio-11.1/lib/pulseaudio/libpulsecommon-11.1.so)
mkdir: /nix/store/2kcrj1ksd2a14bm5sky182fv2xwfhfap-glibc-2.26-131/lib/libc.so.6: version `GLIBC_2.27' not found (required by /nix/store/zajwxkz8al80j8vrmfv8kacbsbs09rxw-systemd-238-lib/lib/libsystemd.so.0)
```

**Comment by numinit** [comment]
Gone hunting for a root cause, this looks like it:

https://github.com/svenstaro/dwarf_fortress_unfuck/commit/c25627ae92eee04b694a54fd57cf976d8fdd807c#diff-af3b638bc2a3e6c650974192a53c7291R4

libgraphics in unfuck is now built in release mode, which is removing some symbols dfhack needs.

**Comment by numinit** [comment]
This should be mostly working now, people can checkout numinit/nixpkgs:df if they want a working 0.44.11+therapist+dfhack+unfuck.

Still some weird things going on with therapist not finding memory layout files. Had to manually symlink the therapist share directory into my home folder. Will look into that more.

Would also like a nicer fix for dfhack breaking on Linux, but patching CMakeLists.txt seems to work for now.

So, with all that considered, would like to make some more changes before squashing and requesting a review, but this is a starting point for any of itching to try 0.44.11 on NixOS.

**Comment by matthewbauer** [comment]
Ok pushed a commit to fix a few issues. Hope everything still works okay.

**Comment by numinit** [comment]
Looks like unfuck got an update upstream. I've been running the fixed version locally; will submit another pull later.

Thanks for the feedback on my first NixOS pull :-)

**Comment by Baughn** [comment]
This updates to an unstable alpha version of dfhack, which warns that it's likely to corrupt game data structures. I know it's called `nixos-unstable`, but still...

**Comment by matthewbauer** [comment]
Hmmm... should we have stayed at the old dfhack? The issue is that dfhack is kind of worthless if it doesn't work on the dwarf fortress available.

**Comment by matthewbauer** [comment]
We actually could provide older versions of dwarf fortress fairly easily with some .json. Maybe that would lessen the burden in updating dwarf fortress to make everything work at once?

**Comment by Baughn** [comment]
Considering https://github.com/NixOS/nixpkgs/pull/43081, I'd very much like us to provide a works-by-default "lazy" package. If that means holding back DF, as it likely does, then... yes, I think we're going to need that.

**Pull request merged**